src/k2/persistence/plog_client/PlogClient.cpp

```diff
@@ -0,0 +1,182 @@
+    }
+
+    return seastar::when_all_succeed(appendFutures.begin(), appendFutures.end())
+        .then([this, expected_offset, appended_offset](std::vector<std::tuple<Status, dto::PlogAppendResponse> >&& results) { 
+            Status return_status;
+            for (auto& result: results){
+                auto& [status, response] = result;
+                return_status = std::move(status);
+                if (!return_status.is2xxOK()) 
+                    break;
+                if (response.offset != expected_offset || response.bytes_appended != appended_offset){
+                    return_status = Statuses::S500_Internal_Server_Error("offset inconsistent");
+                    break;
+                }
+            }
+            return seastar::make_ready_future<std::tuple<Status, uint32_t> >(std::tuple<Status, uint32_t>(std::move(return_status), std::move(expected_offset)));
+        });
+}
+
+
+seastar::future<std::tuple<Status, Payload>> PlogClient::read(String plogId, uint32_t offset, uint32_t size){
+    dto::PlogReadRequest request{.plogId = std::move(plogId), .offset=offset, .size=size};
+
+    return RPC().callRPC<dto::PlogReadRequest, dto::PlogReadResponse>(dto::Verbs::K23SI_PERSISTENT_READ, request, *_partitionMapEndpoints[_partitionNameList[_partition_map_pointer]][0], plog_request_timeout()).
+        then([this] (auto&& result) {
+            auto& [status, response] = result;
+
+            return seastar::make_ready_future<std::tuple<Status, Payload> >(std::tuple<Status, Payload>(std::move(status), std::move(response.payload)));
+        });
+}
+
+seastar::future<std::tuple<Status, uint32_t>> PlogClient::seal(String plogId, uint32_t offset){
+    dto::PlogSealRequest request{.plogId = std::move(plogId), .offset=offset};
+
+    std::vector<seastar::future<std::tuple<Status, dto::PlogSealResponse> > > sealFutures;
+    for (auto& ep:_partitionMapEndpoints[_partitionNameList[_partition_map_pointer]]){
+        sealFutures.push_back(RPC().callRPC<dto::PlogSealRequest, dto::PlogSealResponse>(dto::Verbs::K23SI_PERSISTENT_SEAL, request, *ep, plog_request_timeout()));
+    }
+
+    return seastar::when_all_succeed(sealFutures.begin(), sealFutures.end())
+        .then([this](std::vector<std::tuple<Status, dto::PlogSealResponse> >&& results) { 
+            Status return_status;
+            uint32_t sealed_offset;
+            for (auto& result: results){
+                auto& [status, response] = result;
+                return_status = std::move(status);
+                sealed_offset = response.offset;
+                if (!return_status.is2xxOK()) 
+                    break;
+            }
+            return seastar::make_ready_future<std::tuple<Status, uint32_t> >(std::tuple<Status, uint32_t>(std::move(return_status), std::move(sealed_offset)));
+        });
+}
+
+String PlogClient::generate_plogId(){
+    std::string appendix;

```

Comment:
I think it'd be more efficient if you start with a string that is big enough to hold the prefix and the numeric code and then just shuffle the last chars
https://en.cppreference.com/w/cpp/algorithm/random_shuffle
``` c++
    String plogid = "TPCC_CLIENT_plog_0123456789";
    std::mt19937 g(rd());
    std::shuffle(plogid.begin()+18, plogid.end(), g);
```

Reply:
fixed